Handle adding and removing deps in AssetsDefinition.map_asset_specs

dpeng817 opened the pull request:
# Background
This PR changes `AssetsDefinition.map_asset_specs` to allow adding additional deps to an `AssetSpec`.

To do this, we create `Nothing` inputs under the hood which map to the asset dependency in question.

This is important mostly for integration cases where we use the decorator pattern `@dbt_assets`, `@fivetran_assets`, etc. If you want to link an asset dependency to assets defined with those decorators, it's currently not feasible for users.

# Caveats

**Only supports op-backed assets**
This PR adds support only for op-backed assets, not graph-backed assets. I attempted to add support and found myself blocked since graph-backed assets currently do not support non-argument deps (at least from what I could gather). I added a linear issue to track this and referenced it in the relevant place in code.

**Doesn't work for argument-based deps**
This PR adds support for the deletion of non-argument based deps; but does not support the deletion of argument-based deps. I think deletion of argument-based deps is impossible, since you would have nothing to pass to the underlying function. We error cleanly instead, and I think that's the best we can do.

# Testing

A bunch of tests for every case that I could think of:
- add additional dependencies
- add the same dependency multiple times
- remove a dependency
- add dependencies with partition mappings
- add dependencies to a non-executable assets def
- attempt to add deps to a graph-backed asset

dpeng817 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26144** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26141** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26141?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    if not assets_def.is_executable:
+        return assets_def
```
why the early return?

dpeng817 replied:
there's no op if the asset isn't executable, and attempting to access the property throws an error - so we shouldn't be attempting any op-level munging.

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1286,27 +1290,32 @@ def update_replace_dict_and_conflicts(
+            for dep in spec.deps:
+                if dep.asset_key not in deps_per_key:
+                    raise DagsterInvalidDefinitionError(
+                        f"After mapping, asset {dep.asset_key.to_user_string()} was not found in the mapped dependencies. You should not remove dependencies from assets, since that can destroy the internal structure of the asset."
```
hmmmm why not let the user do this? I'm not really following why this should always be illegal

schrockn replied:
Hmmm that does not match up to the description in the error message. If this is for  arguemnt-based deps that should be explained in the errror message. However this should work for vanilla deps.

dpeng817 replied:
I'll have to dig into whether there's a way to disambiguate argument based deps from non-argument based deps.

dpeng817 replied:
I was able to find a way to disambiguate, so I think we're good to support this 👍

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+        return assets_def
+    if not isinstance(assets_def.node_def, OpDefinition):
+        raise DagsterInvalidDefinitionError("Cannot add dependencies to a graph-backed asset")
+    additional_input_names_by_key = {
```
super confusing variable name since it is actually indexed by name

dpeng817 replied:
oof. Yea I'll fix that

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -85,6 +85,10 @@
 ASSET_SUBSET_INPUT_PREFIX = "__subset_input__"
 
 
+def stringify_asset_key_to_input_name(asset_key: AssetKey) -> str:
```
is there not an existing utility that does this? We should use a single, canonical one

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+    }
+
+    additional_ins = {
+        input_name: In(dagster_type=Nothing) for input_name in additional_input_names_by_key
```
given the name/key inversion this is basically unreviewable so will just comment on the idea

schrockn reviewed (dismissed):
It makes sense to support this.

One point of discussion is the behavior of `merge_attributes` w/r/t to deps. My initial thought is to make it work like the `tags` and `metadata` where collission are just overwritten. Essentially treat it like a dictionary indexed by asset key.

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1910,74 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+    )
+    if not isinstance(assets_def.node_def, OpDefinition):
+        raise DagsterInvalidDefinitionError("Cannot add dependencies to a graph-backed asset")
+    dep_key_by_input_name = {
```
nit: generally like blank lines after dedents in Python for readability

schrockn commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1910,74 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
     return non_secure_md5_hash_str(json.dumps(sorted_key_strs).encode("utf-8"))[:8]
+
+
+def _thread_deps_to_node(
```
function name a bit add.

`_assets_def_with_additional_deps` or something like that more clear

dpeng817 replied:
changed the name, also should reflect the fact that we're potentially deleting deps now too

schrockn reviewed (changes requested):
Ok this is looking better.

A few things

* Remove the RFC tag from the PR title if this is meant for final review
* Rewrite the summary to be worthy of a real commit. Correct, proper English, etc.
* We can't just not do `graph_asset` because " I don't think support would be that hard to add, just involves a bunch of additional gross munging that I don't really want to do, and graph-backed assets have limited adoption anyway."

dpeng817 commented:
Okay, now that I've validated direction and incorporated feedback, I think this PR is ready for final review. I think I've addressed all outstanding comments.

dpeng817 commented:
@schrockn ; given your current increased responsibilities and given that I've addressed all outstanding comments, I'll have @OwenKephart review this one to the finish line.

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1896,108 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+        return assets_def
+    original_key_to_input_mapping = reverse_dict(assets_def.node_keys_by_input_name)
+    # Get the new deps that are not already in the original mapping
+    new_deps = {
```
This code is a bit hard to follow, which is partly unavoidable as we have all these annoying dictionaries, but the way I'm thinking about it, it might be easier to handle this purely as sets to start with, i.e.:

```python
# mapped_specs is the list of specs that were mapped in the above fn
new_deps = set().union(*(spec.deps for spec in mapped_specs))
previous_deps = set().union(*(spec.deps for spec in original_assets_def.specs))

added_deps = new_deps-previous_deps
deleted_deps = previous_deps-new_deps

if not added_deps and not deleted_deps: ...
```

Possible that I'm missing some subtle or not-so-subtle stuff that makes that pattern not work, so if that's the case feel free to ignore

dpeng817 replied:
quality suggestion. I think the core complication comes from me trying to save a linear pass over the replaced specs. Which is .... dum

OwenKephart reviewed (changes requested):
agree this should be added, have some comments on code structure here

dpeng817 commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1896,108 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+    )
+
+    updated_dep_keys = {key: {dep.asset_key for dep in deps} for key, deps in updated_deps.items()}
+    return AssetsDefinition(
```
AssetsDefinition.dagster_internal_init explicitly doesn't allow you to pass both attributes dict _and_ specs, which is why we need to do the double-assetsdefinition-construction thing here.

OwenKephart replied:
not quite following this -- the existing code was already calling `self.__class__.dagster_internal_init(**{**self.get_attributes_dict(), "specs": mapped_specs})`, and the only difference here seems to be that we're changing `keys_by_input_name`, which is a valid parameter for `dagster_internal_init`. So what goes wrong when you do:

```
self.__class__.dagster_internal_init(
    **{**self.get_attributes_dict(), "specs": mapped_specs, "keys_by_input_name": ...}
)
```
?

OwenKephart replied:
anyway the concern here (beyond it being verbose) is that this is sorta a ticking timebomb if we ever add an additional parameter to AssetsDefinition and forget to add it here

dpeng817 replied:
Ah - turns out I was wrong. I was under the impression that you could _not_ set both input-to-asset-key mappings and specs, but it turns out you can.

dpeng817 commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1286,27 +1290,22 @@ def update_replace_dict_and_conflicts(
             mapped_specs.append(mapped_spec)
 
-        return self.__class__.dagster_internal_init(
+        assets_def_new_specs = self.__class__.dagster_internal_init(
```
I can take a pass at restructuring. The reason I did this way was to avoid re-building the asset dependency structure, but perhaps that's actually unavoidable anyway. I do so much dictionary munging anyway I'm not convinced I've actually saved any time complexity

OwenKephart commented on python_modules/dagster/dagster/_core/definitions/assets.py:
```diff
@@ -1897,3 +1895,64 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
+    remaining_original_deps = previous_deps - removed_deps
+    original_key_to_input_mapping = reverse_dict(assets_def.node_keys_by_input_name)
+
+    # If there are no changes to the dependency structure, return the original assets_def
```
nit: slightly confusing comment as we're not returning the original assets_def (we're returning an assets def with the original node)

dpeng817 replied:
oof you right

dpeng817 replied:
fixed

OwenKephart reviewed (approved):
excellent
